### ISSUE TITLE
[SURE-9393] Rancher crash looping (panic) after provisioning an EKS cluster

### ISSUE THREAD

**Issue by kkaempf** [rancher/rancher#48144]
### Steps to reproduce:

Provision a new EKS cluster with all default settings. Under "Networking" choose "Select from existing subnets" and select two existing subnets from a pre-existing VPC.
Wait for the cluster to provision. 

Result:

After some time Rancher will crash with a panic.

The following error is logged in the Rancher pods (see attached log file):
 
~~~
E1119 21:15:45.659899 39 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference) goroutine 2549 [running]: k8s.io/apimachinery/pkg/util/runtime.logPanic({0x82a1c60, 0x100f0fa0}) /go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:75 +0x85 k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc009370750?}) /go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:49 +0x6b panic({0x82a1c60?, 0x100f0fa0?}) /usr/lib64/go/1.22/src/runtime/panic.go:770 +0x132 github.com/rancher/rancher/pkg/controllers/management/eks.(*eksOperatorController).onClusterChange(0xc00461f500, {0x9c9d9e0?, 0xc0209b1b01?}, 0x46222d?) /app/pkg/controllers/management/eks/eks_cluster_handler.go:204 +0x8f8
~~~
 
 The Rancher pods will enter crash looping state and never recover again.

I exported the eksclusterconfig object for the provisioned cluster (see attachments).

----
<details><summary>Rancher Pod Logs</summary>
<p>

```
2024/11/20 09:34:45 [INFO] Starting cluster controllers for local
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider freeipa because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider okta because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider keycloak because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider activedirectory because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider oidc because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider ping because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
2024/11/20 09:34:45 [INFO] Refusing to reset the config and clean up resources of the auth provider github because its auth config annotation management.cattle.io/auth-provider-cleanup is set to rancher-locked.
E1120 09:34:45.456487      39 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 2033 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x82a1c60, 0x100f0fa0})
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc01576f300?})
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x82a1c60?, 0x100f0fa0?})
	/usr/lib64/go/1.22/src/runtime/panic.go:770 +0x132
github.com/rancher/rancher/pkg/controllers/management/eks.(*eksOperatorController).onClusterChange(0xc004a470a0, {0x9c9d9e0?, 0xc0056cbb01?}, 0x46222d?)
	/app/pkg/controllers/management/eks/eks_cluster_handler.go:204 +0x8f8
github.com/rancher/wrangler/v3/pkg/generic.(*Controller[...].func1({0xaed9238?, 0xc002e83508?})
	/go/pkg/mod/github.com/rancher/wrangler/v3@v3.0.0/pkg/generic/controller.go:169 +0x44
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0x9c9d9e0?, {0xc0030cc6c6?, 0x9e1999d?}, {0xaed9238?, 0xc002e83508?})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/sharedcontroller.go:29 +0x32
github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc00143e0f0, {0xc0030cc6c6, 0x7}, {0xaed9238, 0xc0030d0a88})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/sharedhandler.go:75 +0x202
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc001074790, {0xc0030cc6c6, 0x7})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:236 +0x12e
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc001074790, {0x79d6560, 0xc01576f300})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:217 +0xeb
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc001074790)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:194 +0x45
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:183
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001209b60, {0xaec5e20, 0xc0021b2900}, 0x1, 0xc0010fc000)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001209b60, 0x3b9aca00, 0x0, 0x1, 0xc0010fc000)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:161
created by github.com/rancher/lasso/pkg/controller.(*controller).run in goroutine 1539
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:151 +0x2ba
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xe0 pc=0x6de8038]

goroutine 2033 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc01576f300?})
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:56 +0xcd
panic({0x82a1c60?, 0x100f0fa0?})
	/usr/lib64/go/1.22/src/runtime/panic.go:770 +0x132
github.com/rancher/rancher/pkg/controllers/management/eks.(*eksOperatorController).onClusterChange(0xc004a470a0, {0x9c9d9e0?, 0xc0056cbb01?}, 0x46222d?)
	/app/pkg/controllers/management/eks/eks_cluster_handler.go:204 +0x8f8
github.com/rancher/wrangler/v3/pkg/generic.(*Controller[...].func1({0xaed9238?, 0xc002e83508?})
	/go/pkg/mod/github.com/rancher/wrangler/v3@v3.0.0/pkg/generic/controller.go:169 +0x44
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0x9c9d9e0?, {0xc0030cc6c6?, 0x9e1999d?}, {0xaed9238?, 0xc002e83508?})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/sharedcontroller.go:29 +0x32
github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc00143e0f0, {0xc0030cc6c6, 0x7}, {0xaed9238, 0xc0030d0a88})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/sharedhandler.go:75 +0x202
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc001074790, {0xc0030cc6c6, 0x7})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:236 +0x12e
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc001074790, {0x79d6560, 0xc01576f300})
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:217 +0xeb
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc001074790)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:194 +0x45
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:183
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001209b60, {0xaec5e20, 0xc0021b2900}, 0x1, 0xc0010fc000)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001209b60, 0x3b9aca00, 0x0, 0x1, 0xc0010fc000)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:161
created by github.com/rancher/lasso/pkg/controller.(*controller).run in goroutine 1539
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20240923125127-ae858d002589/pkg/controller/controller.go:151 +0x2ba
```

</p>
</details>

**Comment by thehejik** [comment]
Maybe we should not require `DescribeAddon` permission when EBS or any other Addon is NOT requested. On the other hand rancher shouldn't crash when the permission is added afterwards.

**Comment by yiannistri** [comment]
Rancher runs a set of controllers, two of which are relevant to this issue:
- `eks-operator-controller`
- `cluster-refresher-controller`

The `eks-operator-controller`  is used to set fields and conditions that the `eks-operator` itself does not handle (i.e. set the cluster driver to EKS, set the managed launch template versions in status etc.).

The `cluster-refresher-controller` is used to refresh the upstream spec of a cluster, effectively to reflect any changes that happened upstream.

Both of these controllers at some point need to set the status  from upstream. This means that they need to reach out to the EKS API and retrieve the cluster and its settings as it exists on AWS. Part of upstream status is also whether the EBS CSI addon is enabled or not. If the account used to make the API call lacks the EKS:DescribeAddon permission, then the upstream status retrieved is incomplete (instead of checking whether it's enabled or not it will fail with an error) and the operator will not make the right call.

Therefore in my opinion, the EKS:DescribeAddon permission should be part of the minimum EKS permissions and it doesn't matter whether the addon is enabled or not (additional permissions are needed to enable it/disable it) as it's used to get the upstream status.

**Comment by valaparthvi** [comment]
>To conclude, I think the best thing to do is to make sure the EKS:DescribeAddon permission is mentioned in the docs and is part of the minimum EKS permissions. I don't think we should worry about the scenario when a permission is added/removed halfway through cluster creation.

Rancher Docs does mention [minimum required permission](https://ranchermanager.docs.rancher.com/v2.10/how-to-guides/new-user-guides/kubernetes-clusters-in-rancher-setup/set-up-clusters-from-hosted-kubernetes-providers/eks#minimum-eks-permissions) which includes EBS permissions, so we are good there.

But I was wondering if we should check and ensure UpstreamSpec to be non-empty before using any of it's field, and return an error if it is empty.

**Comment by thehejik** [comment]
The fix has been successfully verified as working on `rancher:v2.9-36ad76da660f93b0f71245487385829eabe4956f-head` with `rancher/eks-operator:v1.9.4`.

After spotting the permission error and adding the `eks:DescribeAddon` permission additionally to used IAM policy the EKS cluster is provisioned successfully and its state is `Active` after a while. The rancher pod doesn't crash and reporting following error:
```
2024/12/06 11:18:24 [WARNING] failed to set initial upstreamSpec on cluster [c-c97xp]: error checking if ebs csi driver addon is installed: operation error EKS: DescribeAddon, https response error StatusCode: 403, RequestID: bf7dc27c-a753-46d4-bf1a-54f31b5b0e29, api error AccessDeniedException: User is not authorized to perform this action
```

Next step is to backport the fix to `release/v2.10` and `main` branches.

**Comment by thehejik** [comment]
**Test report** 
Verified on:
* `rancher:v2.9-b8239f00fb45d7bfcfc31b2b6a8d7c8a20f755bf-head` with `rancher/eks-operator:v1.9.5-rc.1`.
* `rancher:v2.10-49414bccf043599e1d3b0910a496e765d5945baf-head` with `rancher/eks-operator:1.10.1-rc.2`

On both versions the rancher pod doesn't crash anymore but the cluster is not resurrected after adding the permission. On AWS I can see the cluster is `Active` and also its NodePool is up.

The original error message disappears after adding the permission but then the state turns to `Waiting` with message `Waiting for API to be available` and it will stay like that forever (well I was waiting for ~30mins).
![image](https://github.com/user-attachments/assets/c25fe46f-4f2c-4bbe-8b8d-4c3859c7cabb)

Faulty EKS cluster needs to be recreated once again, unfortunately Edit also doesn't lead to recovery.

Anyway the goal of this issue is probably met. The pod doesn't crash anymore and rancher is still fully operational.

---
This is strip of logs from eks-operator:
```
time="2024-12-10T16:55:59Z" level=error msg="error syncing 'cattle-global-data/c-4z2s7': handler eks-controller: error checking if ebs csi driver addon is installed: operation error EKS: DescribeAddon, https response error StatusCode: 403, RequestID: 9a3bed60-17ec-4ae9-b1ea-7dedeacb853a, api error AccessDeniedException: User is not authorized to perform this action, requeuing"
time="2024-12-10T16:56:33Z" level=info msg="Waiting for cluster [th-eks (id: c-4z2s7)] to update nodegroups [group1]"
time="2024-12-10T16:56:35Z" level=info msg="Waiting for cluster [th-eks (id: c-4z2s7)] to update nodegroups [group1]"
time="2024-12-10T16:57:06Z" level=info msg="Waiting for cluster [th-eks (id: c-4z2s7)] to update nodegroups [group1]"
time="2024-12-10T16:57:37Z" level=info msg="Waiting for cluster [th-eks (id: c-4z2s7)] to update nodegroups [group1]"
time="2024-12-10T16:58:08Z" level=info msg="Waiting for cluster [th-eks (id: c-4z2s7)] to update nodegroups [group1]"
time="2024-12-10T16:58:40Z" level=info msg="Cluster [th-eks (id: c-4z2s7)] finished updating"
```
[Logs from rancher pod](https://github.com/user-attachments/files/18083317/rancher-58c97ff5df-8lk82_rancher.log)

**Comment by thehejik** [comment]
This issue is fixed, rancher doesn't crash anymore on 2.9 and 2.10. Also the documentation has been adapted to require `esk:DescribeAddon` permission.

Closing in favor of follow up issue https://github.com/rancher/rancher/issues/48366
